contracts/CompoundMiddleContract.sol

```diff
@@ -187,125 +119,182 @@ contract CompoundMiddleContract {
+     * @param _erc20Address address of erc20 token contract
+     * @param _comptrollerAddress address of comptroller contract in Compound
+     * @param _amountToBorrow amount of erc20 tokens to borrow
+     * @return uint256 borrowBalance of the user
+     */
+    function borrowErc20WithEth(
+        address payable _cEtherAddress,
+        address _erc20Address,
         address _comptrollerAddress,
         address _cTokenAddress,
-        uint256 _underlyingDecimals
-    ) external payable returns (bool) {
+        uint256 _amountToBorrow
+    ) external payable returns (uint256) {
         // Create references to Compound and Token contracts
-        
+        CEth cEth = CEth(_cEtherAddress);
+        CErc20 cToken = CErc20(_cTokenAddress);
+        IERC20 token = IERC20(_erc20Address);
+        Comptroller comptroller = Comptroller(_comptrollerAddress);
 
         // Deposit Eth as collateral
+        require(cEth.balanceOf(address(this)) > 0, "DEPOSIT ETHER FIRST");
 
         // enter market with Eth
+        address[] memory cTokens = new address[](1);
+        cTokens[0] = _cEtherAddress;
+        uint256[] memory errors = comptroller.enterMarkets(cTokens);
+        require(errors[0] == 0, "Comptroller.enterMarkets FAILED");
+
+        (uint256 error2, uint256 liquidity, uint256 shortfall) = comptroller.getAccountLiquidity(address(this));
+        require(error2 == 0, "comptroller.getAccountLiquidity FAILED");
+        require(shortfall == 0, "account underwater");
+        require(liquidity > 0, "account has excess collateral");
+
+        console.log("Liquidity available: ", liquidity);
 
         // borrow
+        uint256 borrowStatus = cToken.borrow(_amountToBorrow);
+        require(borrowStatus == 0, "BORROW FAILED");
 
         // get borrow balance
+        uint256 borrowBalance = cToken.borrowBalanceCurrent(address(this));
+
+        // transfer borrowed erc20 to user
+        token.safeTransfer(owner, _amountToBorrow);
+
+        return borrowBalance;
+    }
+
+    /**
+     * @dev repays erc20 tokens to Compound
+     * @param _cErc20Address address of cErc20 contract in Compound
+     * @param _erc20Address address of the erc20 token contract
+     * @param _repayAmount number of erc20 tokens to be repayed
+     * @return uint256 updated borrowBalance for the erc20 token
+     */
+    function paybackErc20(
+        address _cErc20Address,
+        address _erc20Address,
+        uint256 _repayAmount
+    ) external returns (uint256) {
+        // create references to contracts
+        CErc20 cToken = CErc20(_cErc20Address);
+        IERC20 token = IERC20(_erc20Address);
+
+        // transfer user's tokens to contract
+        token.safeTransferFrom(msg.sender, address(this), _repayAmount);
+
+        // approve Compound to spend erc20 tokens
+        token.safeApprove(_cErc20Address, _repayAmount);
+
+        // repay borrow
+        cToken.repayBorrow(_repayAmount);
+
```

Comment:
Are you sure in here, that repayBorrow succeeds in all cases?

Reply:
handled